sway-ir/src/pass_manager.rs

```diff
@@ -120,41 +161,93 @@ impl PassManager {
     /// Run the passes specified in `config`.
     pub fn run(&mut self, ir: &mut Context, config: &PassManagerConfig) -> Result<bool, IrError> {
         let mut modified = false;
-        for pass in &config.to_run {
-            let pass_t = self.passes.get(pass.as_str()).expect("Unregistered pass");
+        let mut worklist: Vec<_> = config
+            .to_run
+            .iter()
+            .rev()
+            .map(|pass| self.passes.get(pass).expect("Unregistered pass"))
+            .collect();
+        while !worklist.is_empty() {
```

Comment:
> I found this a bit complicated to understand, maybe split it the code into an helper or two or maybe even split it up into different phases? Like one phase to calculate the necessary pass ordering, and a second to execute them.

I've rewritten that algorithm now to use recursion rather than iteration+worklist. It's much simpler now. Recursion should be safe since the recursion depth (worst case) won't go beyond the total number of static passes we have in the compiler.